sunpy/net/dataretriever/sources/norh.py

```diff
@@ -35,119 +34,55 @@ class NoRHClient(GenericClient):
-
-        Parameters
-        ----------
-        timerange: `sunpy.time.TimeRange`
-            time range for which data is to be downloaded.
-
-        Returns
-        -------
-        urls : list
-            list of URLs corresponding to the requested time range
-        """
-
-        # We allow queries with no Wavelength but error here so that the query
-        # does not get passed to VSO and spit out garbage.
-        if 'wavelength' not in kwargs.keys() or not kwargs['wavelength']:
-            raise ValueError("Queries to NORH should specify either 17GHz or 34GHz as a Wavelength."
-                             "see https://solar.nro.nao.ac.jp/norh/doc/manuale/node65.html")
-        else:
-            wavelength = kwargs['wavelength']
-
-        # If wavelength is a single value GenericClient will have made it a
-        # Quantity in the kwargs.
-        if not isinstance(wavelength, u.Quantity):
-            raise ValueError(f"Wavelength to NORH must be one value not {wavelength}.")
-
-        wavelength = wavelength.to(u.GHz, equivalencies=u.spectral())
-        if wavelength == 34 * u.GHz:
-            freq = 'tcz'
-        elif wavelength == 17 * u.GHz:
-            freq = 'tca'
-        else:
-            raise ValueError("NORH Data can be downloaded for 17GHz or 34GHz,"
-                             " see https://solar.nro.nao.ac.jp/norh/doc/manuale/node65.html")
-
-        # If start of time range is before 00:00, converted to such, so
-        # files of the requested time ranger are included.
-        # This is done because the archive contains daily files.
-        if timerange.start.strftime('%M-%S') != '00-00':
-            timerange = TimeRange(timerange.start.strftime('%Y-%m-%d'),
-                                  timerange.end)
-
-        norh = Scraper(BASEURL, freq=freq)
-        # TODO: warn user that some files may have not been listed, like for example:
-        #       tca160504_224657 on ftp://solar-pub.nao.ac.jp/pub/nsro/norh/data/tcx/2016/05/
-        #       as it doesn't follow pattern.
-
-        return norh.filelist(timerange)
-
-    def _get_time_for_url(self, urls):
-        freq = urls[0].split('/')[-1][0:3]  # extract the frequency label
-        crawler = Scraper(BASEURL, freq=freq)
-        times = list()
-        for url in urls:
-            t0 = crawler._extractDateURL(url)
-            # hard coded full day as that's the normal.
-            times.append(TimeRange(t0, t0 + TimeDelta(1*u.day)))
-        return times
-
-    def _makeimap(self):
-        """
-        Helper Function used to hold information about source.
-        """
-        self.map_['source'] = 'NAOJ'
-        self.map_['provider'] = 'NRO'
-        self.map_['instrument'] = 'NORH'
-        self.map_['physobs'] = ''
+    baseurl = r'ftp://solar-pub.nao.ac.jp/pub/nsro/norh/data/tcx/%Y/%m/(\w){3}%y%m%d'
+    pattern = '{}/tcx/{year:4d}/{month:2d}/{Wavelength:3l}{:4d}{day:2d}'
+    optional = {a.Wavelength, a.Source, a.Provider}
```

Comment:
I thought the plan was to have `required` as a class attribute and derive optional from `register_attrs`?

Reply:
This can't be done, since we can't derive `wavelength` from register_attrs as we planned not to register it.